python/ml4ir/applications/ranking/model/metrics/metrics_helper.py

```diff
@@ -3,16 +3,131 @@
 
 
+def compute_failure_stats(
+    df: pd.DataFrame,
+    query_key_col: str,
+    label_col: str,
+    old_rank_col: str,
+    new_rank_col: str,
+    group_keys: List[str] = [],
+    secondary_labels: List[str] = [],
+):
+    """Compute failure metrics using secondary labels"""
+
+    def compute_failure_stats_group(query_group):
+        old_click_rank = query_group[query_group[label_col] == 1][old_rank_col].values[0]
+        new_click_rank = query_group[query_group[label_col] == 1][new_rank_col].values[0]
+
+        failure_metrics_dict = {
+            k: v[0] for k, v in query_group[group_keys].to_dict(orient="list").items()
+        }
+        for secondary_label in secondary_labels:
+            click_secondary_label = query_group[query_group[label_col] == 1][
+                secondary_label
+            ].values[0]
+
+            old_pre_click_secondary_label = query_group[
+                query_group[old_rank_col] < old_click_rank
+            ][secondary_label]
+            new_pre_click_secondary_label = query_group[
+                query_group[new_rank_col] < new_click_rank
+            ][secondary_label]
+
+            old_failure_all = 0
+            old_failure_any = 0
+            old_failure_count = 0
+            old_failure_fraction = 0
+            if old_pre_click_secondary_label.size > 0:
+                # Query failure only if failure on all records
+                old_failure_all = (
+                    1 if (old_pre_click_secondary_label < click_secondary_label).all() else 0
+                )
+                # Query failure if failure on at least one record
+                old_failure_any = (
+                    1 if (old_pre_click_secondary_label < click_secondary_label).any() else 0
+                )
+                # Count of failure records
+                old_failure_count = (old_pre_click_secondary_label < click_secondary_label).sum()
+                # Normalizing to fraction of potential records
+                old_failure_fraction = old_failure_count / (old_click_rank - 1)
+
+            new_failure_all = 0
+            new_failure_any = 0
+            new_failure_count = 0
+            new_failure_fraction = 0
+            if new_pre_click_secondary_label.size > 0:
+                # Query failure only if failure on all records
+                new_failure_all = (
+                    1 if (new_pre_click_secondary_label < click_secondary_label).all() else 0
+                )
+                # Query failure if failure on at least one record
+                new_failure_any = (
+                    1 if (new_pre_click_secondary_label < click_secondary_label).any() else 0
+                )
+                # Count of failure records
+                new_failure_count = (new_pre_click_secondary_label < click_secondary_label).sum()
+                # Normalizing to fraction of potential records
+                new_failure_fraction = new_failure_count / (new_click_rank - 1)
```

Comment:
You could factorise the computation of new & old metrics in a method

Reply:
Yeah. I thought so too. Let me do it.

Reply:
Done.